### ISSUE TITLE
Hashcat with CUDA is broken

### ISSUE THREAD

**Issue by Janrupf** [NixOS/nixpkgs#310234]
### Describe the bug
Attempting to run hashcat with CUDA results in the CUDA stub driver being loaded because `/run/opengl-driver/lib` is not on `LD_LIBRARY_PATH`.

### Steps To Reproduce
1. enable `nixpkgs.config.cudaSupport = true`
2. use the hashcat package
3. run hashcat

### Expected behavior
`/run/opengl-driver/lib` should be added to the `LD_LIBRARY_PATH` automatically so that hashcat runs.

### Screenshots
```console
> hashcat hash.txt
hashcat (6.2.6) starting in autodetect mode

cuDriverGetVersion(): 34

Started: Thu May  9 01:59:22 2024
Stopped: Thu May  9 01:59:22 2024
```
![Screenshot of the terminal output above](https://img.janrupf.net/i/4bko2.png)

### Additional context
The package seems to try to add the OpenGL run path, but it doesn't work (for some reason?).

See here:
https://github.com/NixOS/nixpkgs/blob/9a35766bd30c07f39c881fc439db65ee1fff65b2/pkgs/tools/security/hashcat/default.nix#L71-L76

### Notify maintainers
@felixalbrigtsen @kierdavis @zimbatm

### Metadata
```console
> nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.29, NixOS, 24.05 (Uakari), 24.05.20240502.63c3a29`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/p69bcs7ma6ijj8v9xsrg3nq3nn8ryn95-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by felixalbrigtsen** [comment]
I believe it is the same problem that is seen and explained in  https://github.com/NixOS/nixpkgs/issues/272221#issuecomment-1848816792, and agree that only the stub driver is loaded. 
My computer with proper Nvidia drivers is unavailable at the moment, so I can't test it out right now, but I have some ideas:

Have you enabled the graphics drivers like described at https://nixos.wiki/wiki/Nvidia (or similar documentation for your type of graphics card), and does it seem to work in other applications that use the GPU? 

Does it work if you hard code that library path, something like the diff below?

```diff
diff --git a/pkgs/tools/security/hashcat/default.nix b/pkgs/tools/security/hashcat/default.nix
index 56399e61639e..9a239c0b7e3a 100644
--- a/pkgs/tools/security/hashcat/default.nix
+++ b/pkgs/tools/security/hashcat/default.nix
@@ -64,6 +64,7 @@ stdenv.mkDerivation rec {
       "${ocl-icd}/lib"
     ] ++ lib.optionals cudaSupport [
       "${cudaPackages.cudatoolkit}/lib"
+      "/run/opengl-driver/lib"
     ]);
   in ''
     wrapProgram $out/bin/hashcat \
```

**Comment by Janrupf** [comment]
Nope, that doesn't fix the issue - but almost. Changing the order in which the paths are added does actually fix the issue, so this works:
```nix
ib.optionals cudaSupport [
  "/run/opengl-driver/lib" # Note that this needs to be first
  "${cudaPackages.cudatoolkit}/lib"
]);
```

**Comment by felixalbrigtsen** [comment]
Thank you for a good report!

From the definition of `addOpenGLRunpath`, I think the current hashcat package _should_ have worked as is, but I am not really sure if it's a bug or if I got something wrong.
It looks like at least some other applications run`addOpenGLRunpath` on the shared libraries: https://github.com/NixOS/nixpkgs/blob/47b604b07d1e8146d5398b42d3306fdebd343986/pkgs/development/python-modules/tensorflow/default.nix#L565-L569 We only run it on the one executable, but our hashcat libraries in `$out/share` don't seem to reference libcuda directly(?).

Alternatively, some programs, like https://github.com/NixOS/nixpkgs/blob/e5a9c059efe2fae982ffdd555d9b7ae677fde76f/nixos/modules/services/misc/plex.nix#L179 just put opengl-drivers in the LD_LIBRARY_PATH like we did in the comments above. Maybe it would be okay to simply make change you tested permanent?
I don't think I see a problem with "hard coding" `/run/opengl-driver/lib` like this.

**Comment by eclairevoyant** [comment]
fyi `addDriverRunpath` is the current preference over `addOpenGLRunpath`

**Comment by SomeoneSerge** [comment]
As a reminder, the fix is to add `${getLib cuda_cudart}/lib` to DT_RUNPATH.

RE: "revert getOutput"
The change was reintroduced in https://github.com/NixOS/nixpkgs/pull/323056

**Comment by jr1221** [comment]
> As a reminder, the fix is to add `${getLib cuda_cudart}/lib` to DT_RUNPATH.
> 
> RE: "revert getOutput" The change was reintroduced in #323056
If you dont mind me asking where should that be placed in the hashcat `default.nix` file?  Is it an additon to the wrapper?

**Comment by SomeoneSerge** [comment]
You can just add `autoAddDriverRunpath` to `nativeBuildInputs`
